### PR TITLE
Enable github annotations parsable format

### PR DESCRIPTION
**Pull request by ssbarnea**
Detects if is run as a github action/workflow and adapts its output
accordingly.

### CODE REVIEW

File: lib/ansiblelint/formatters/__init__.py [hunk 1]
```diff
@@ -87,6 +87,26 @@ def format(self, match, colored=False):
...
+    ::warning file={name},line={line},col={col}::{message}
+    """
+
+    def format(self, match, colored=False):
+        formatstr = u"::warning file={0},line={1}::[{2}] {3}"
+        if colored:
+            return formatstr.format(colorize(self._format_path(match.filename), Color.filename),
+                                    colorize(str(match.linenumber), Color.linenumber),
+                                    colorize(u"E{0}".format(match.rule.id), Color.error_code),
+                                    colorize(u"{0}".format(match.message), Color.error_title))
```
Comment (webknjaz) [inline lib/ansiblelint/formatters/__init__.py:102]: I'm pretty sure that colorizing will break this feature.

File: lib/ansiblelint/formatters/__init__.py [hunk 2]
```diff
@@ -87,6 +87,26 @@ def format(self, match, colored=False):
                                     match.message)
 
 
+class AnnotationsFormatter(BaseFormatter):
+    """Emits errors in format that is recognized by Github Annotation parsing.
+
+    ::warning file={name},line={line},col={col}::{message}
+    """
+
+    def format(self, match, colored=False):
```
Comment (webknjaz) [inline lib/ansiblelint/formatters/__init__.py:96]: This should probably print both to console and to workflow commands. Because the log must still remain readable.

File: lib/ansiblelint/__main__.py [hunk 3]
```diff
@@ -71,7 +72,10 @@ def main() -> int:
         formatter_factory = formatters.QuietFormatter
 
     if options.parseable:
-        formatter_factory = formatters.ParseableFormatter
+        if os.environ.get('GITHUB_WORKFLOW'):
```
Comment (webknjaz) [inline lib/ansiblelint/__main__.py:75]:
```suggestion
        if os.getenv('GITHUB_WORKFLOW'):
```

File: lib/ansiblelint/formatters/__init__.py [hunk 4]
```diff
@@ -87,6 +87,20 @@ def format(self, match, colored=False):
...
+        formatstr = u"::warning file={0},line={1}::[{2}] {3}"
```
Comment (webknjaz) [inline lib/ansiblelint/formatters/__init__.py:97]: I think this shouldn't be only `warning`. Can we use `error` based on severity?

File: README.rst [hunk 5]
```diff
@@ -117,6 +117,15 @@ The following is the output from ``ansible-lint --help``, providing an overview
...
 
+CI/CD
+-----
+
+If execution under `Github Actions`_ is detected via the presence of
+``GITHUB_WORKFLOW`` variable, the linter will also print errors using their
+`annotation`_ format.
+
+.. _GitHub Actions: https://github.com/features/actions
+.. _annotation: https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-error-message
```
Comment (webknjaz) [inline README.rst:128]: Here's how you can match the length of the paragraph above
```suggestion
.. _annotation:
   https://docs.github.com/en/actions/reference
   /workflow-commands-for-github-actions#setting-an-error-message
```

File: lib/ansiblelint/__main__.py [hunk 6]
```diff
@@ -118,6 +119,12 @@ def main() -> int:
     for match in sorted(matches):
         print(formatter.format(match, options.colored))
 
+    # If run under GitHub Actions we also want to emit output recognized by it.
+    if os.getenv('GITHUB_WORKFLOW'):
```
Comment (webknjaz) [inline lib/ansiblelint/__main__.py:123]: This doc https://docs.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables#default-environment-variables suggests that `GITHUB_ACTIONS` var will always equal `true` in such envs. So let's rely on this property to avoid false positives in some artificial test envs:
```suggestion
    if os.getenv('GITHUB_ACTIONS') == 'true' and os.getenv('GITHUB_WORKFLOW'):
```

File: lib/ansiblelint/formatters/__init__.py [hunk 7]
```diff
@@ -15,7 +15,7 @@ class BaseFormatter:
         display_relative_path (bool): whether to show path as relative or absolute
     """
 
-    def __init__(self, base_dir, display_relative_path):
+    def __init__(self, base_dir=str, display_relative_path=True) -> None:
```
Comment (webknjaz) [inline lib/ansiblelint/formatters/__init__.py:18]: typo?
```suggestion
    def __init__(self, base_dir: str, display_relative_path=True) -> None:
```

File: lib/ansiblelint/formatters/__init__.py [hunk 8]
```diff
@@ -87,6 +87,33 @@ def format(self, match, colored=False):
                                     match.message)
 
 
+class AnnotationsFormatter(BaseFormatter):
+    # https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-a-warning-message
+    """Emits errors in format that is recognized by Github Annotation parsing.
+
+    ::warning file={name},line={line},col={col}::{message}
```
Comment (webknjaz) [inline lib/ansiblelint/formatters/__init__.py:94]: How about mentioning all the possibilities?
```suggestion
    ::debug file={name},line={line},col={col}::{message}
    ::error file={name},line={line},col={col}::{message}
    ::warning file={name},line={line},col={col}::{message}
```

File: lib/ansiblelint/formatters/__init__.py [hunk 9]
```diff
@@ -87,6 +87,33 @@ def format(self, match, colored=False):
...
+    def format(self, match, colored=False):
+        formatstr = u"::{0} file={1},line={2}::[{3}] {4}"
+        return formatstr.format(
+            self.severity_to_level(match.rule.severity),
+            self._format_path(match.filename),
+            match.linenumber,
+            "E" + match.rule.id,
+            match.message)
+
+    def severity_to_level(self, severity: str) -> str:
```
Comment (webknjaz) [inline lib/ansiblelint/formatters/__init__.py:108]: I think this shouldn't be embedded into the class (it does nothing with the instance. But if you want to keep it, turn into a `@staticmethod` at least, and add a leading underscore.

File: lib/ansiblelint/formatters/__init__.py [hunk 10]
```diff
@@ -87,6 +87,33 @@ def format(self, match, colored=False):
...
+    """
+
+    def format(self, match, colored=False):
+        formatstr = u"::{0} file={1},line={2}::[{3}] {4}"
+        return formatstr.format(
+            self.severity_to_level(match.rule.severity),
+            self._format_path(match.filename),
+            match.linenumber,
+            "E" + match.rule.id,
+            match.message)
```
Comment (webknjaz) [inline lib/ansiblelint/formatters/__init__.py:106]: Let's use named placeholders, numbers in templates are confusing.
```suggestion
        workflow_command = self.severity_to_level(match.rule.severity)
        file_path = self._format_path(match.filename)
        line_num = match.linenumber
        rule_id = match.rule.id
        violation_details = match.message

        return (
            f"::{workflow_command} file={file_path},line={line_num}"
            f"::[E{rule_id}] {violation_details}"
        )
```

File: lib/ansiblelint/formatters/__init__.py [hunk 11]
```diff
@@ -87,6 +87,33 @@ def format(self, match, colored=False):
...
+class AnnotationsFormatter(BaseFormatter):
+    # https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-a-warning-message
+    """Emits errors in format that is recognized by Github Annotation parsing.
+
+    ::warning file={name},line={line},col={col}::{message}
+
+    Supported levels: debug, warning, error
+    """
+
+    def format(self, match, colored=False):
```
Comment (webknjaz) [inline lib/ansiblelint/formatters/__init__.py:99]: Let's explicitly state the invocation contract:
```suggestion
    def format(self, match, colored=False):
        """Prepare a match instance for reporting as a GitHub Actions annotation."""
        if colored:
            raise ValueError('The colored mode is not supported.')

```

File: README.rst (hunk 5 above)
Comment (ssbarnea) [reply README.rst:128]: If our linters jobs is not complaining lets not go into this. URLs are known to cause problems
when wrapped and I do not want to risk.

File: tox.ini [hunk 12]
```diff
@@ -31,6 +31,7 @@ deps =
 commands =
   # safety measure to assure we do not accidentaly run tests with broken dependencies
   python -m pip check
+  -ansible-lint -p examples/example.yml
```
Comment (webknjaz) [inline tox.ini:34]: don't forget to remove this before merging
```suggestion
```

File: lib/ansiblelint/__main__.py [hunk 13]
```diff
@@ -117,6 +118,12 @@ def main() -> int:
     for match in sorted(set(matches)):
         print(formatter.format(match, options.colored))
 
+    # If run under GitHub Actions we also want to emit output recognized by it.
+    if os.getenv('GITHUB_ACTIONS') == 'true' and os.getenv('GITHUB_WORKFLOW'):
+        formatter = formatters.AnnotationsFormatter(cwd, options.display_relative_path)
```
Comment (webknjaz) [inline lib/ansiblelint/__main__.py:123]: This should always use a relative path. Otherwise, the annotations won't be mapped to things in the project:
```suggestion
        formatter = formatters.AnnotationsFormatter(cwd, True)
```

File: tox.ini [hunk 14]
```diff
@@ -42,6 +42,7 @@ install_command =
     {packages}
 passenv =
   CURL_CA_BUNDLE  # https proxies, https://github.com/tox-dev/tox/issues/1437
+  GITHUB_*
```
Comment (webknjaz) [inline tox.ini:45]: I don't think this belongs here. We don't really rely on it in tests atm.

**Comment by webknjaz** [comment]
@ssbarnea I'd like some tests for this new formatter. Also, we need tests for the default auto-selection inside and outside of GHA (probably parametrized + `monkeypatch.setenv()`).

**Comment by ssbarnea** [comment]
> @ssbarnea I'd like some tests for this new formatter. Also, we need tests for the default auto-selection inside and outside of GHA (probably parametrized + `monkeypatch.setenv()`).

I added test, so we would also cover this condition.

File: lib/ansiblelint/formatters/__init__.py [hunk 15]
```diff
@@ -15,7 +16,7 @@ class BaseFormatter:
         display_relative_path (bool): whether to show path as relative or absolute
     """
 
-    def __init__(self, base_dir, display_relative_path):
+    def __init__(self, base_dir: Union[str, Path], display_relative_path=True) -> None:
```
Comment (webknjaz) [inline lib/ansiblelint/formatters/__init__.py:19]: Why does this need to default to `True`?

File: lib/ansiblelint/__main__.py [hunk 16]
```diff
@@ -117,6 +118,12 @@ def main() -> int:
     for match in sorted(set(matches)):
```
Comment (webknjaz) [inline lib/ansiblelint/__main__.py:118]: @ssbarnea this is `sorted(set())` and the one below is just `sorted()`.
1) I think that both should be `sorted(set())`
2) Also, it'd be better to just safe this in a variable and reuse it in two places

File: lib/ansiblelint/formatters/__init__.py [hunk 17]
```diff
@@ -90,6 +91,42 @@ def format(self, match, colored=False):
                                     match.message)
 
 
+class AnnotationsFormatter(BaseFormatter):
+    # https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-a-warning-message
+    """Emits errors in format that is recognized by Github Annotation parsing.
```
Comment (webknjaz) [inline lib/ansiblelint/formatters/__init__.py:96]: This mustn't start with a verb because classes represent objects and don't do anything by themselves, the actions are in methods and methods should have imperative strings.

```suggestion
    """Formatter for emitting violations as GitHub Workflow Commands.

    These commands trigger the GHA Workflow runners platform to post violations
    in a form of GitHub Checks API annotations that appear rendered in pull-
    request files view.
```

File: test/conftest.py [hunk 18]
```diff
@@ -19,4 +21,19 @@ def default_text_runner(default_rules_collection):
     return RunFromText(default_rules_collection)
 
 
-# vim: et:sw=4:syntax=python:ts=4:
```
Comment (webknjaz) [inline test/conftest.py:22]: Keep this.

File: test/conftest.py [hunk 19]
```diff
@@ -19,4 +21,19 @@ def default_text_runner(default_rules_collection):
...
+def run_ansible_lint(cwd, role_path=None, bin=None, env=None):
```
Comment (webknjaz) [inline test/conftest.py:24]: Conftest is not the best location for such helpers. It should have pytest-related things like fixtures and hooks. Put this into `__init__.py` along with other helpers instead of having them spread across multiple locations w/o any structure.

Also, this change is unrelated to GHA annotations and must be submitted separately.

File: test/TestCliRolePaths.py [hunk 20]
```diff
@@ -90,12 +73,30 @@ def test_run_playbook(self):
...
+
+def test_run_playbook_github():
+    """Call ansible-lint simulating GitHub Actions environment."""
+    cwd = os.path.dirname(os.path.realpath(__file__))
+    role_path = '../examples/example.yml'
+
+    env = {
+        "GITHUB_ACTIONS": "true",
+        "GITHUB_WORKFLOW": "foo"
+    }
```
Comment (webknjaz) [inline test/TestCliRolePaths.py:95]: We need a negative test where these values don't match the condition. You could use parameterized to keep them all in one test function.

File: test/TestCliRolePaths.py [hunk 21]
```diff
@@ -90,12 +73,30 @@ def test_run_playbook(self):
...
+
+    result = run_ansible_lint(cwd=cwd, role_path=role_path, env=env)
+    expected = (
+        '::error file=../examples/example.yml,line=47::[E101] '
```
Comment (webknjaz) [inline test/TestCliRolePaths.py:99]: It is bad if the formatter really emits paths like this, with `../`. They should be resolved and relative to the git repo root. Otherwise, GHA won't match the annotations with files in the repo.

Comment (ssbarnea) [reply test/TestCliRolePaths.py:99]: Not really true. By design ansilble lint emits relative paths relative to CWD when command was called. Same behavior can be seen from other linters too. The "git" repo part is defined for all use-cases.

This path is like this because we call it from inside `./test`, but we can easily avoid it. Still, is not a bug.

Comment (webknjaz) [reply test/TestCliRolePaths.py:99]: It is a bug because it breaks the main purpose of the feature.

**Comment by ssbarnea** [comment]
Closing in favor of #926 but I will address the test comment.
